### ISSUE TITLE
compute encodings seems to make quantized model worse?

### ISSUE THREAD

**Issue by Winfriedvdd** [quic/aimet#1705]
Hi! I am not sure if I correctly understands what happens, but when I start with a pretrained model, and do the following:

quantsim = QuantizationSimModel(model=model, quant_scheme = 'tf_enhanced', ...)
eval_result_before=evaluate(quantsim.model, ...)
quantsim.compute_encodings(...)
eval_result_after=evaluate(quantsim.model,...)

I find that eval_result_after is usually worse than eval_result_before...
Are we not supposed to use quantsim.model before having computed encodings? And if so, is there any other way to tell how much the results have improved due to computing_encodings?

Kind regards,

Winfried

**Comment by LLNLanLeN** [comment]
@Winfriedvdd hey (just another fellow aimet user), I believe that before doing `quantsim.compute_encodings`, the model itself is still a float model, without any quantization parameters (you can verify this by `print(quantsim)` and I believe all of the quantization parameters will be None). 

But, when you run `quantsim.compute_encodings(...)`, and provide X amounts of put, the function will now calculate all the quantization statistic (such as max, min, offset, etc...) and then applied to the model (you should be able to check by print out the model again). Now the model contains int8 weights (float weight to int8 weight model conversation based on the calculated statistic). If you want to improve the model after `compute_encodings`, you can either use post-training quantization or QAT and you should see improvement after that. Hope that helps

**Comment by quic-mangal** [comment]
@LLNLanLeN, thanks for answering the question. 
@Winfriedvdd, let us know if you have any further questions.

**Comment by quic-akhobare** [comment]
@Winfriedvdd - giving some insights to your original question - 

> is there any other way to tell how much the results have improved due to computing_encodings?

So there is a chicken and egg issue. To try any quantized eval, we need some encodings. So there is no way to get an accuracy score without computing encodings. What you can try for sure if to try different quantization schemes when creating the QuantizationSimModel which use different schemes to compute encodings. And compare the accuracy between these different schemes.

**Comment by Winfriedvdd** [comment]
Thanks for the clear answer!

I did notice very small differences compared to the original floating point implementation, so I assumed something must have happened, but now I am unsure what then exactly changed. Not that it matters all too much, just curious as `print(quantsim)` indeed shows all quantizers are disabled.

Perhaps unrelated, but I was wondering if there is any incentive (other than that it takes some time perhaps) to not use large amounts (or all) of the training data for the `quantsim.compute_encodings` step. For example, if we plan on using QAT afterwards, is there a risk of getting stuck in some local minimum making the QAT afterwards less fruitful?

I also noticed that setting bw to 32 is significantly worse than expected when it comes to evaluation loss after the `quantsim.compute_encodings` step. I do see a gradual decrease in performance when making bitwidths smaller moving from 30 to 4, but at 32 something happens that I cannot really place, and the model performs even worse than the bw 4 version...

Kind regards,

Winfried

**Comment by ShobithaShivakumar** [comment]
This is actually expected behavior in the v1 API. When bitwidth=32 is set, AIMET skips encoding computation entirely and sets encodings to None - effectively disabling quantization for those tensors. So bw=32 should produce results identical to the original float model, not a "32-bit quantized" model. If you saw degraded accuracy at bw=32, it may have been due to some quantizers still being active (e.g., if only weight or only activation bw was set to 32) or other configuration issues. If both default_output_bw=32 and default_param_bw=32, the result should match float. 

@Winfriedvdd please re-open this issue if you are still hitting issues with the latest aimet-torch (2.26). If you have further questions, feel free to join our http://aihub.qualcomm.com/community/slack specifically the #aimet channel to discuss easily!
